source/gfoidl.DataCompression/Compression/DeadBandCompression.cs

```diff
@@ -62,114 +65,296 @@ public DeadBandCompression(double instrumentPrecision, in TimeSpan maxTime)
+                    case 1:
+                        TList source      = _source;
+                        int snapShotIndex = _snapShotIndex;
+                        int incomingIndex = _incomingIndex;
 
-            (double Min, double Max) bounding = this.GetBounding(snapShot);
+                        while (true)
+                        {
+                            // Actually a while loop, but so the range check can be eliminated
+                            // https://github.com/dotnet/coreclr/issues/15476
+                            if ((uint)incomingIndex >= (uint)source.Count || (uint)snapShotIndex >= (uint)source.Count)
+                                break;
 
-            for (int i = 1; i < data.Count; ++i)
-            {
-                incoming    = data[i];
-                var archive = this.IsPointToArchive(lastArchived, bounding, incoming);
+                            ref var archive = ref this.IsPointToArchive(incomingIndex);
 
-                if (!archive.Archive)
-                {
-                    snapShot = incoming;
-                    continue;
-                }
+                            if (!archive.Archive)
+                            {
+                                snapShotIndex = incomingIndex++;
+                                continue;
+                            }
 
-                if (!archive.MaxDelta)
-                    yield return snapShot;
+                            if (!archive.MaxDelta)
+                            {
+                                _current = source[snapShotIndex];
+                                _state = 2;
+                                _snapShotIndex = snapShotIndex;
+                                _incomingIndex = incomingIndex;
+                                return true;
+                            }
 
-                yield return incoming;
+                            _snapShotIndex = snapShotIndex;
+                            _incomingIndex = incomingIndex;
+                            goto case 2;
+                        }
 
-                this.UpdatePoints(ref snapShot, ref lastArchived, incoming, archive.MaxDelta, ref bounding);
+                        _current = source[incomingIndex - 1];
+                        _state   = -1;
+                        return true;
+                    case 2:
+                        _current = _source[_incomingIndex];
+                        _state   = 1;
+                        this.UpdatePoints();
+                        _incomingIndex++;
+                        return true;
+                    case DisposedState:
+                        ThrowHelper.ThrowIfDisposed(nameof(DataPointIterator));
+                        return false;
+                }
             }
+            //-----------------------------------------------------------------
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            private void GetBounding(int snapShotIndex)
```

Comment:
Maybe this can be combined with `EnumerableIterator.GetBounding` -- i.e. move it to the base class.